### PR TITLE
Generalization of Test Suite

### PR DESCRIPTION
**Pull request by ansermino**
Initially I set out to add tests for the 128/256bit types. I quickly realized that the static test suite was going to be laborious to maintain and enforce standardization. As a solution I have rewritten the suite such that new types can be added without the need for addition tests to be written.

See `tests/TESTING.md` for details of which tests are currently included and how to add new types.

**At this time a number of tests are failing. As this project is not currently in use I see it as beneficial to (eventually) close this PR and open issues to address these failures so other contributors might lend a hand**

### CODE REVIEW

**Comment by ansermino** [comment]
Closes #1

**Comment by GregTheGreek** [comment]
> Closes #1

I believe you need to add this to the commit message for github to auto close the issue

File: tests/Autotest-config.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    },
+    {
+        name: "Int256",
+        constructor: Int256,
+        flag: "_int256",
+        size: 256,
+    },
+];
+
+export const getInts = (): IntTypeTestInterface[]  => {
```
Comment (GregTheGreek) [inline tests/Autotest-config.ts:148]: +1 for being really clever

File: tests/Autotest-Uint-SafeMath.test.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import {DivisionByZeroError} from "../lib/errors";
+import {assertErrorMsg, validateUint} from "./helpers";
+import {getUints} from "./Autotest-config";
+
+getUints().forEach((tc) => {
+    describe(`SafeMath tests for ${tc.name}`, () => {
+        it(`should return a ${tc.name} for all basic methods: Add`, () => {
+            const one: any = tc.constructor("1");
```
Comment (GregTheGreek) [inline tests/Autotest-Uint-SafeMath.test.ts:8]: Could we replace type `any` with `Uint`?

File: tests/Autotest-Int-SafeMath.test.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
+import {DivisionByZeroError} from "../lib/errors";
+import {assertErrorMsg, validateInt} from "./helpers";
+import {getInts} from "./Autotest-config";
+
+getInts().forEach((tc) => {
+    describe(`SafeMath Tests for ${tc.name}`, () => {
+        it(`should return an ${tc.name} for all basic methods: Add`, () => {
+            const one: any = tc.constructor("1");
```
Comment (GregTheGreek) [inline tests/Autotest-Int-SafeMath.test.ts:8]: Same with Uint-safeMath, can we use type `int` vs `any`

File: tests/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        fn();
+    } catch (e) {
+        err = e;
+    }
+    if (!err) {
+        expect.fail("No error thrown");
+    } else {
+        expect(err.message).to.equal(expected.message);
+    }
+};
```
Comment (GregTheGreek) [inline tests/helpers.ts:17]: The `if/else` doesn't need to be after,`try/catch` will handle all of that logic for you. The following should work.
```

export const assertErrorMsg = (fn , expected: Error) => {
    try {
        fn();
        // If fn() throws, this line will not be reachable
        expect.fail("No error thrown");
    } catch (e) {
        expect(err.message).to.equal(expected.message);
    }
};
```

**Review by GregTheGreek — CHANGES_REQUESTED** [review]
Please read comments. Otherwise looks really great, I like this style and we should adopt this wherever possible.
Perhaps update the CI with this PR, and we can close #10 first, then #9 after.

File: tests/Autotest-Uint-SafeMath.test.ts (hunk 2 above)
Comment (ansermino) [reply tests/Autotest-Uint-SafeMath.test.ts]: Typescript wasn't happy with the Uint/Int super-types. I think that was my initial intent behind #9, having identical constructors I think resolves this, will pull from master and see if it lets me.

Comment (ansermino) [reply tests/Autotest-Uint-SafeMath.test.ts:8]: I get the error `TS2349: Cannot invoke an expression whose type lacks a call signature` when using Int/Uint. Any ideas? @JonathanLorimer?

Comment (JonathanLorimer) [reply tests/Autotest-Uint-SafeMath.test.ts:8]: @ansermino I will have to look more in depth at this. But I had two ideas about what might be the issue. The first is that you might be overriding the functions call signature, see this [post](https://stackoverflow.com/questions/39691889/error-cannot-invoke-an-expression-whose-type-lacks-a-call-signature) a possible fix if this is the case would be: 
```javascript 
const two = one.add(one) as Uint;
```
My second suspicion is that, due to the use of generics to get an exact type returned by the math methods, there is an issue when you use an alias that could possibly be multiple types. The constructors all partially apply an expected return type to all the math methods. So the expected type to be returned (for a Uint8 for example) is exactly Uint8, not Uint8 | Uint16 | Uint32 ... etc. A potential solution for this is to add a big case switch, but that kind of defeats the purpose of generalizing the tests. If this is the issue then I suggest we go with any type. There are a couple situations in the constructors where I had to use any types for this exact reason.

Comment (ansermino) [reply tests/Autotest-Uint-SafeMath.test.ts:8]: After many attempts I think I'm going to propose we put this off for now. This seems to be an additional feature and does not impose any immediate limitations as far as I can see. I will open an issue to address it and we can revisit it with any bright ideas that may come about.

File: tests/helpers.ts (hunk 4 above)
Comment (ansermino) [reply tests/helpers.ts:17]: Excellent point, will update

**Review by GregTheGreek — APPROVED** [review]

**Review by GregTheGreek — CHANGES_REQUESTED** [review]
CI stil failing

**Review by GregTheGreek — APPROVED** [review]

**Pull request merged**